src/external/song.ts

```diff
@@ -17,100 +17,186 @@
 export type SongAnalysis = {
-  analysisId: string;
-  analysisState: string;
-  studyId: string;
-  firstPublishedAt?: string;
-  [k: string]: any;
+	analysisId: string;
+	analysisState: string;
+	studyId: string;
+	firstPublishedAt?: string;
+	[k: string]: unknown;
+};
+
+type SongResponseAnalysesPage = {
+	analyses: SongAnalysis[];
+	totalAnalyses: number;
+	currentTotalAnalyses: number;
+};
+
+type PaginationInputs = {
+	limit: number;
+	offset: number;
+};
+
+type PaginationMetadata = PaginationInputs & {
+	responseCount: number;
+	totalCount: number;
 };
 
 export const getStudies = async (url: string) => {
-  const studiesUrl = urljoin(url, '/studies/all');
-  logger.info(`Fetching studies from: ${studiesUrl}`);
-  const res = await fetch(`${url}/studies/all`);
-  const studies = await res.json();
-  logger.info(`Retrieved ${studies.length} studies: ${studies}`);
-  return studies;
+	const studiesUrl = urljoin(url, '/studies/all');
+	logger.info(`Fetching studies from: ${studiesUrl}`);
+	const res = await fetch(`${url}/studies/all`);
+	const studies = await res.json();
+	logger.info(`Retrieved ${studies.length} studies: ${studies}`);
+	return studies;
 };
 
-/**
- * Fetch potentially very large responses from Song then return them in a batched stream.
- * @param analysesUrl
- * @returns
- */
-const fetchAnalysesInBatches = async (analysesUrl: string): Promise<Batch> => {
-  const controller = new abortController();
-  const timeoutPeriod = (await getAppConfig()).datacenter.fetchTimeout;
-  const batchSize = (await getAppConfig()).datacenter.batchSize;
-  const timeout = setTimeout(() => {
-    logger.warn(`Aborting request for analyses due to timeout after ${timeoutPeriod}`);
-    controller.abort();
-  }, timeoutPeriod);
-  try {
-    logger.info(`Fetching analyses from ${analysesUrl}`);
-    const res = await fetch(analysesUrl, {
-      signal: controller.signal,
-    });
-    const resStream = res.body;
-    const pipeline = resStream.pipe(streamArray.withParser()).pipe(new Batch({ batchSize }));
-    logger.info(`Received analyses response, returning as Batch.`);
-    return pipeline;
-  } catch (error) {
-    throw error;
-  } finally {
-    clearTimeout(timeout);
-  }
+const fetchAnalysesPage = async (inputs: {
+	dataCenter: DataCenter;
+	studyId: string;
+	offset: number;
+	limit: number;
+	analysisStates?: Set<SongAnalysisState>;
+}): Promise<{ data: SongAnalysis[] } & PaginationMetadata> => {
+	try {
+		const {
+			dataCenter: { centerId, songUrl },
+			studyId,
+			limit,
+			offset,
+			analysisStates,
+		} = inputs;
+		const analysisStatesQueryValue = analysisStates
+			? Array.from(analysisStates).join(',')
+			: SongAnalysisStates.PUBLISHED || SongAnalysisStates.PUBLISHED;
+
+		const queryParams = {
+			analysisStates: analysisStatesQueryValue,
+			limit,
+			offset,
+		};
+
+		const queryString = qs.stringify(queryParams);
+		const analysesUrl = urljoin(songUrl, '/studies', studyId, 'analysis', `paginated?${queryString}`);
```

Comment:
non-functional suggestion, feel free to dismiss: query params in urlJoin are idiomatically expressed like so.

```suggestion
		const analysesUrl = urljoin(songUrl, '/studies', studyId, 'analysis', 'paginated', `?${queryString}`);
```

Reply:
Let me just double check that `paginated?stuff=things` and `paginated/?stuff=things` work the same way for this service.

Reply:
Works fine 👍